docs/solidity-by-example.rst

```diff
@@ -645,4 +645,499 @@ Safe Remote Purchase
+
+Here's the modified **javascript code** to cryptographic a message from the previous chapter:
+
+::
+
+    function constructPaymentMessage(contractAddress, amount) {
+        return ethereumjs.ABI.soliditySHA3(
+            ["address", "uint256"],
+            [contractAddress, amount]
+        );
+    }
+    
+    function signMessage(message, callback) {
+        web3.personal.sign(
+            "0x" + message.toString("hex"),
+            web3.eth.defaultAccount,
+            callback
+        );
+    }
+    
+    // contractAddress is used to prevent cross-contract replay attacks.
+    // amount, in wei, specifies how much ether should be sent.
+    
+    function signPayment(contractAddress, amount, callback) {
+        var message = constructPaymentMessage(contractAddress, amount);
+        signMessage(message, callback);
+    }
+
+Closing the Payment Channel
+---------------------------
+
+When the recipient is ready to receive their funds, it's time to
+close the payment channel by calling a *close* function on the smart contract.
+Closing the channel pays the recipient the ether they're owed and destroys the contract,
+sending any remaining ether back to the sender.
+To close the channel, the recipient needs to share a message signed by the sender.
+
+The smart contract must verify that the message contains a valid signature from the sender.
+The process for doing this verification is the same as the process the recipient uses.
+The Solidity functions *isValidSignature* and *recoverSigner* work just like their
+JavaScript counterparts in the previous section. The latter is borrowed from the
+*ReceiverPays* contract in the previous chapter.
+
+The *close* function can only be called by the payment channel recipient,
+who will naturally pass the most recent payment message because that message
+carries the highest total owed. If the sender were allowed to call this function,
+they could provide a message with a lower amount and cheat the recipient out of what they're owed.
+
+The function verifies the signed message matches the given parameters.
+If everything checks out, the recipient is sent their portion of the ether,
+and the sender is sent the rest via a *selfdestruct*.
+You can see the *close* function in the full contract.
+
+Channel Expiration
+-------------------
+
+The recipient can close the payment channel at any time, but if they fail to do so,
+the sender needs a way to recover their escrowed funds. An *expiration* time was set
+at the time of contract deployment. Once that time is reached, the sender can call
+*claimTimeout* to recover their funds. You can see the *claimTimeout* function in the
+full contract.
+
+After this function is called, the recipient can no longer receive any ether,
+so it's important that the recipient close the channel before the expiration is reached.
+
+
+The full contract
+-----------------
+
+::
+
+    pragma solidity ^0.4.20;
```

Comment:
IMO, it would be great if we make it for latest compiler.

Reply:
Updating the code right now!